### DISCUSSION TITLE
better detection of broken profiles

### DISCUSSION THREAD

**Discussion by jcupitt** [libvips/libvips#2084] [Ideas]
This image has an embedded profile that is no longer supported by lcms2:

https://t0.nl/IEC61966-2.1.jpg

(discovered by @kleisauke)

lcms reports that there is no matrix shaper, and also no clut, so there's not much it can do with it. git master libvips does this:

```
john@yingna ~/pics $ vipsthumbnail IEC61966-2.1.jpg -e srgb -o x.jpg -n perceptual

(vipsthumbnail:3582973): VIPS-WARNING **: 13:49:09.096: icc_transform: intent 0 (perceptual) not supported by input profile; falling back to default intent
vipsthumbnail: unable to thumbnail IEC61966-2.1.jpg
VipsIcc: Couldn't link the profiles
```

You can't override the broken input profile either:

```
john@yingna ~/pics $ vipsthumbnail IEC61966-2.1.jpg -i srgb -e srgb -o x.jpg -n perceptual

(vipsthumbnail:3764987): VIPS-WARNING **: 14:21:04.422: icc_transform: intent 0 (perceptual) not supported by input profile; falling back to default intent
vipsthumbnail: unable to thumbnail IEC61966-2.1.jpg
VipsIcc: Couldn't link the profiles
```

**Comment by jcupitt** [comment]
This branch adds more profile checks during load, so the bad profile is rejected and fallback can happen.

https://github.com/libvips/libvips/tree/improve-detection-of-bad-profiles

With this change, you now see: 

```
john@yingna ~/pics $ vipsthumbnail IEC61966-2.1.jpg -e srgb -o x.jpg -n perceptual

(vipsthumbnail:3766011): VIPS-WARNING **: 14:23:44.137: input profile does not support perceptual intent
vipsthumbnail: unable to thumbnail IEC61966-2.1.jpg
icc_transform: no input profile
```

If you supply a fallback input profile, you get:

```
john@yingna ~/pics $ vipsthumbnail IEC61966-2.1.jpg -i srgb -e srgb -o x.jpg -n perceptual

(vipsthumbnail:3765975): VIPS-WARNING **: 14:23:15.906: input profile does not support perceptual intent
```

A warning, but the thumbnail is generated.

**Comment by jcupitt** [comment]
Issues: 

- Having to supply a fallback profile is bad, we should fall back to srgb / cmyk / mono etc. automatically
- The error message could be better

**Comment by kleisauke** [comment]
I can confirm that on the [`improve-detection-of-bad-profiles`](https://github.com/libvips/libvips/tree/improve-detection-of-bad-profiles) branch the fallback input profile is honored with this particular image. Thanks!
